### PR TITLE
Adds a pipeline for creating reader's output

### PR DESCRIPTION
**Pull request by MadVikingGod**
This patch adds a pipeline for creating and managing readers' outputs.  

Each reader should have a unique pipeline registered the reader, and when creating an instrument you should add aggregators for each view the instrument is in.  

This structure is intended to be used as part of a larger Registry structure.  That is in a separate patch to minimize the scope of this PR.

The use of a new aggregator interface is to avoid coloring this struct.  If the `addAggregator()` took an `internal.Aggregator[N]` then I wouldn't be able to have int and float aggregators in the same slice.

### CODE REVIEW

File: sdk/metric/pipeline.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	if p.aggregations[scope] == nil {
+		p.aggregations[scope] = map[instrumentKey]aggregator{}
+	}
+	inst := instrumentKey{
+		name:        name,
+		description: description,
+		unit:        instUnit,
+	}
+	if _, ok := p.aggregations[scope][inst]; ok {
+		return fmt.Errorf("instrument already registered: name %s, scope: %s", name, scope)
```
Comment (jmacd) [inline sdk/metric/pipeline.go:80]: Since this is part of a larger construction, I can't quite answer the following: What happens next after this error happens? Is this treated somewhere or returned to the user?

**Review by dashpole — APPROVED** [review]

File: sdk/metric/pipeline.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+// produce returns aggregated metrics from a single collection.
+//
+// This method is safe to call concurrently.
+func (p *pipeline) produce(ctx context.Context) (metricdata.ResourceMetrics, error) {
+	p.Lock()
+	defer p.Unlock()
+
+	for _, callback := range p.callbacks {
+		// TODO make the callbacks parallel.
```
Comment (MrAlias) [inline sdk/metric/pipeline.go:102]: Can we add an issue to track this?

File: sdk/metric/provider.go [hunk 3]
```diff
@@ -53,9 +57,20 @@ func NewMeterProvider(options ...Option) *MeterProvider {
...
+	}
+
 	return &MeterProvider{
-		res:        conf.res,
-		readers:    conf.readers,
+		res:       conf.res,
+		readers:   conf.readers,
+		providers: providers,
+
+		meters: meterRegistry{},
```
Comment (MrAlias) [inline sdk/metric/provider.go:72]:
```suggestion
```

File: sdk/metric/pipeline.go (hunk 1 above)
Comment (MrAlias) [reply sdk/metric/pipeline.go:80]: @MadVikingGod correct me if I'm wrong, but I imagine this is going to be returned from the instrument creation methods (i.e. https://github.com/open-telemetry/opentelemetry-go/blob/096a1624184285edbf9c82c53ea0b068895915ef/metric/instrument/asyncfloat64/asyncfloat64.go#L24-L34)

File: sdk/metric/pipeline.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	}
+	if _, ok := p.aggregations[scope][inst]; ok {
+		return fmt.Errorf("instrument already registered: name %s, scope: %s", name, scope)
+	}
+
+	p.aggregations[scope][inst] = agg
+	return nil
+}
+
+// addCallback registers a callback to be run when `produce()` is first called.
```
Comment (MrAlias) [inline sdk/metric/pipeline.go:87]: > addCallback registers a callback to be run when `produce()` is first called.

Is this only run on the first invocation?

File: sdk/metric/provider.go [hunk 5]
```diff
@@ -33,8 +33,12 @@ import (
...
 	res *resource.Resource
 
-	meters  meterRegistry
-	readers map[Reader][]view.View
+	meters    meterRegistry
+	readers   map[Reader][]view.View
+	providers map[Reader]*pipeline
+
+	// callbackMutex sync.Mutex
+	// callbacks map[Reader][]callbacks
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply sdk/metric/provider.go:41]: Right, should this be removed?

File: sdk/metric/pipeline_test.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+	})
+
+	output, err = pipe.produce(context.Background())
+	require.NoError(t, err)
+	assert.Nil(t, output.Resource)
+	assert.Len(t, output.ScopeMetrics, 1)
+}
+
+func TestNewPipeline(t *testing.T) {
+	pipe := newPipeline(nil)
```
Comment (MrAlias) [inline sdk/metric/pipeline_test.go:65]: Missing test where a non-nil resource is passed and ensured to be used by the pipeline.

File: sdk/metric/pipeline_test.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+	err := pipe.addAggregator(instrumentation.Scope{}, "name", "desc", unit.Dimensionless, testSumAggregator{})
+	require.NoError(t, err)
+
+	err = pipe.addAggregator(instrumentation.Scope{}, "name", "desc", unit.Dimensionless, testSumAggregator{})
+	assert.Error(t, err)
+
+	output, err := pipe.produce(context.Background())
+	assert.NoError(t, err)
+	require.Len(t, output.ScopeMetrics, 1)
+	require.Len(t, output.ScopeMetrics[0].Metrics, 1)
```
Comment (MrAlias) [inline sdk/metric/pipeline_test.go:97]: Is there a reason this is tested here and not in the previous tests?

**Review by MrAlias — APPROVED** [review]
Nothing blocking.

File: sdk/metric/pipeline.go (hunk 1 above)
Comment (MadVikingGod) [reply sdk/metric/pipeline.go:80]: Yes, this will eventually bubble up to the instrument creation.  This is the first point where an instrument could clash with an already created instrument.

For example, if you have a reader with two views, everything, and rename foo to bar.  The first instrument (foo or bar) that is registered shouldn't have an error, but the second one will.

File: sdk/metric/pipeline.go (hunk 4 above)
Comment (MadVikingGod) [reply sdk/metric/pipeline.go:87]:
```suggestion
// addCallback registers a callback to be run when `produce()` is called.
```

File: sdk/metric/pipeline.go (hunk 2 above)
Comment (MadVikingGod) [reply sdk/metric/pipeline.go:102]:
```suggestion
		// TODO make the callbacks parallel. ( #3034 )
```

File: sdk/metric/provider.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (MadVikingGod) [reply sdk/metric/provider.go:41]:
```suggestion
```

File: sdk/metric/pipeline_test.go (hunk 6 above)
Comment (MadVikingGod) [reply sdk/metric/pipeline_test.go:67]: [Done](https://github.com/open-telemetry/opentelemetry-go/pull/3026/files#diff-0183e3119b983b08bf970ba61c2e8627312d20f6f5577d06d8466e6a7bc94c0bR103-R109)

**Comment by MrAlias** [comment]
@jmacd are you okay if we merge this? You have the last outstanding comment.
